swarm/src/registry.rs

```diff
@@ -379,7 +378,9 @@ mod tests {
     }
 
     #[test]
+    #[ignore]
     fn score_retention() {
+        return todo!("This test fails randomly");
```

Comment:
Passes for me locally.

Reply:
```
---- registry::tests::score_retention stdout ----
thread 'registry::tests::score_retention' panicked at 'assertion failed: exists', swarm/src/registry.rs:407:22
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'registry::tests::score_retention' panicked at '[quickcheck] TEST FAILED (runtime error). Arguments: (AddressRecord { addr: "/tcp/181", score: Finite(0) }, AddressRecord { addr: "/tcp/67", score: Infinite })
Error: "assertion failed: exists"', /home/kpp/.cargo/registry/src/github.com-1ecc6299db9ec823/quickcheck-1.0.3/src/tester.rs:165:28
```

Reply:
Fails for me now too, maybe a false-positive before.

Reply:
The following patch fixes it:

```patch
Index: swarm/src/registry.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/swarm/src/registry.rs b/swarm/src/registry.rs
--- a/swarm/src/registry.rs	(revision cff5192c82a4f0c2039bba3ea65efa27a8af8694)
+++ b/swarm/src/registry.rs	(date 1662043501670)
@@ -380,11 +380,14 @@
     #[test]
     #[ignore]
     fn score_retention() {
-        return todo!("This test fails randomly");
         fn prop(first: AddressRecord, other: AddressRecord) -> TestResult {
             if first.addr == other.addr {
                 return TestResult::discard();
             }
+
+            if first.score == AddressScore::Finite(0) {
+                return TestResult::discard();
+            }
 
             let mut addresses = Addresses::default();
 
```

Looking at the source, I think what happens is that an `Address` with a score 0 gets added initially but it dropped once _any_ new address is added, regardless of its score. I am not sure if that is the intended behaviour and the test is wrong or vice versa.

Reply:
Okay, so the issue is in first:finite and other:infinite. This test checks this condition:

```
// Only finite scores push out other finite scores.
```

Which is not what we see here. For some reason infinite score pushed out a finite score.

Reply:
> Which is not what we see here. For some reason infinite score pushed out a finite score.

Not really. The zero score pushes it itself out so to speak because zero scores are always removed, regardless of what you add to `Addresses`.

https://github.com/libp2p/rust-libp2p/blob/36a27738618bfa0e90c285f1d379f078d2572730/swarm/src/registry.rs#L204-L215

Reply:
It may be that the earlier version of quickcheck omitted the `0` case for some reason?

Reply:
> It may be that the earlier version of quickcheck omitted the 0 case for some reason?

I ran 100 tests with `assert!(!first.score.is_zero());` and they never fail. Seems this is an unlikely path for an old version. Shall we discard the test on finite=0?

Reply:
Given that we pretty explicitly remove entries where the score is 0, I assume this is an oversight in the test design rather than a bug so yes, I think discarding the input for `score == 0` should be okay.

We might want to add a test that manifests the "remove 0 scores" behaviour.

Reply:
Fixed